R/utils.R

```diff
@@ -386,17 +386,47 @@ serve_book = function(
       if (Rscript(args) != 0) stop('Failed to compile ', paste(files, collapse = ' '))
     }
   }
-  rebuild('index.Rmd', preview_ = FALSE)  # build the whole book initially
+  index <- get_index_file()
+  if (is_empty(index)) {
+    stop("`serve_book()` expect `index.Rmd` or `index.rmd` in the book project.", call. = FALSE)
+  }
+  rebuild(index, preview_ = FALSE)  # build the whole book initially
   servr::httw('.', ..., site.dir = output_dir, handler = rebuild)
 }
 
+index_files <- xfun::with_ext("index", c("Rmd", "rmd"))
+
+get_index_file <- function() {
+  xfun::existing_files(index_files, first = TRUE, error = FALSE)
```

Comment:
To be truly case-insensitive, I recommend that we use `list.files('.', '^index[.]Rmd$', ignore.case = TRUE)` here. That way, filenames like `index.RMD` will also be supported.

Reply:
I did the change. We will just only select the first in alphabetical if there are several index files. 

But one is expected only, so I added a warning really.

Reply:
It also required to check dynamically for index each time we compare for preview
https://github.com/rstudio/bookdown/pull/1351/files#diff-0d68ae244e4a8c5ddc391c48f5399a91f18bc2a955f098e0e66069c607b444f4R213

Should I cache the values of index files now ? 

This was why previously I used `index_files <- xfun::with_ext("index", c("Rmd", "rmd"))`

Reply:
I think `list.files()` should be fast enough, so may not be worth caching. It's up to you.

`list.files(recursive = TRUE)` could be slow, though.